### ISSUE TITLE
Registering a setting with type array causes Schema to not load.

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#2065]
This was reported by @ravewebdev. I did some digging to get to the root of it and explain below. Not sure if it makes sense to anyone but me 🤷🏻‍♂️ , but wanted to get things recorded as I did the digging. 

----

When registering a setting using the WordPress core `register_setting` API and setting the `type` to `array`, the Schema fails to load. 

Take the following snippet:

```php
add_action( 'init', function() {

	register_setting( 'test_setting_group', 'test_setting', [
		'type' => 'array',
		'show_in_graphql' => true
	]);

} );
```

This snippet registers a setting to WordPress, declares the type of setting to be an array, and sets the setting to `show_in_graphql`. 

Since `array` is not a valid GraphQL Type, this **should not** register the field or the Setting Group to the GraphQL Schema. 

However, in v1.5.9 and earlier, this would register an invalid GraphQL Object Type to the schema with no fields. 

We can see this here: 

![Screen Shot 2021-08-09 at 2 14 01 PM](https://user-images.githubusercontent.com/1260765/128768653-d211f9fb-29e5-4dcf-94fd-6e4f63634426.png)

This is a bug and should have never been happening. 

However, since it _was_ happening, some plugins, such as WPEngine Atlas Starter (https://github.com/wpengine/atlas-starter-plugins/blob/main/wpe-atlas-starter/inc/Atlas_Settings.php#L179-L180) were registering invalid objects, then using the `register_graphql_field` API to add fields to the empty object, as seen here: https://github.com/wpengine/atlas-starter-plugins/blob/main/wpe-atlas-starter/inc/Atlas_Settings.php#L282-L284

This behavior _should not_ have been possible, but was allowed (as a bug) because all objects, even invalid ones, were being generated on each request and not being lazily loaded.

When Lazy Loading was fixed with WPGraphQL v1.6, this issue was brought to light, because the `register_graphql_field()` calls here: https://github.com/wpengine/atlas-starter-plugins/blob/main/wpe-atlas-starter/inc/Atlas_Settings.php#L282-L284 are not being applied to any valid GraphQL Object Types that can properly be lazily loaded, and now we're getting errors. 

We can use Introspection to query the Type in WPGraphQL 1.6.2 and see that the Type is indeed invalid. 

WPGraphQL should never have allowed this snippet above to go through with registering a Setting Object Type to the Schema. 

----

Ideas for solutions: 

- WPGraphQL Should do better validation of registered settings to make sure they're registered in a way that can properly map to the GraphQL Schema
- WPGraphQL should provide a `graphql_type` field that can be used to override the WordPress core `type` config for registering settings. For example, register_setting allows `type => 'array'` and `type => 'object'` but these are not valid GraphQL Types. WPGraphQL could allow settings to pass something like `graphql_type => [ 'list_of' => 'String' ]` and `type => 'array'` for use by WordPress core, and they could then play nice together
- WPE Atlas Starter can refactor by setting `show_in_graphql=>false`, then manually register the Settings to the Schema, bypassing WPGraphQLs mapping of registered settings to the Schema.

**Comment by justlevine** [comment]
Am I being short-sighted, or could we simply pass [ `$setting_field['type']` ](https://github.com/wp-graphql/wp-graphql/blob/885196ea28eda45bf9b2d3fabc2b61443b8386f2/src/Type/ObjectType/SettingGroup.php#L69) through a function that replaces the value with a GraphQL compatible type?

Those atlas links are giving me a 404 (private repo?), but if they were able to successfully register fields to whichever invalid  settings object type before 1.6.x, then I'm assuming that means they have a programmatic solve for multidimensional arrays / nested objects for when `['list_of' => 'String']` won't cut it.

**Comment by jasonbahl** [comment]
@justlevine It's hard to explain, but what was happening before 1.6 was that it was adding an empty GraphQL Object Type to the schema. Empty as in an Object Type with no fields. But, since pre-1.6 loaded the FULL schema _always_, it would instantiate the object (always) and call the filter that runs for `register_graphql_field()` which then would add fields to the empty object, making it a valid object. 

That works with the full schema is _always_ loaded. 

But, with Lazy Loading, this is problematic. 

Registering an empty GraphQL Object Type is not allowed and Lazy Loading exposes this flaw. 

Since you can't see the code linked, basically what was happening is:

```php
register_setting( 'settingGroupName', 'settingName' [
  'type' => 'array',
  'show_in_graphql' => true,
]);
```

Then elsewhere in the code was some manual: 

```
register_graphql_field( 'SettingGroupName', 'anotherField', [
  'type' => 'String',
] );
```

What's happening in WPGraphQL is that it picks up the registered settings and tries to build GraphQL Schema from them. 

In this case, we have a Setting Group, named "settingGroupName" which WPGraphQL turns into a GraphQL Object Type named "SettingGroupName", but since the field `settingName` was type `array` the field wasn't added to the GraphQL Object Type, making it an invalid Object Type. WPGraphQL _should_ prevent this, but it's not. 

Without Lazy Loading, the `graphql_$typename_fields` filter would get called when the invalid Object Type was instantiated and the `register_graphql_field( 'SettingGroupName', 'anotherField', ... );` would add a valid field to the object making it _seem_ like everything is fine.

With lazy loading, the `graphql_$typename_fields` only runs when it _needs_ to, which exposes the fact that the Object Type was instantiated empty and causes issues.

**Comment by jasonbahl** [comment]
The initial issue here is resolved by #2262
